src/inmanta/ast/variables.py

```diff
@@ -174,13 +175,17 @@ def get_location(self) -> Location:
         return self.location
 
 
-class IsDefinedReferenceHelper(AbstractAttributeReferenceHelper[bool]):
+class IsDefinedReferenceHelper(AbstractAttributeReferenceHelper[bool], ResultCollector[object]):
     """
     Helper class for IsDefined, reschedules itself
     """
 
     def __init__(self, target: ResultVariable, instance: Optional[Reference], attribute: str) -> None:
         super().__init__(target, instance, attribute, None)
+        self.resultcollector = self
+
+    def receive_result(self, value: T, location: Location) -> None:
+        self.target.set_value(True, self.location)
 
     def target_value(self) -> bool:
```

Comment:
This will be called even if the value is already set? (Which is OK in itself, but not strictly required)

Reply:
In my initial implementation I'd overridden `resume` to do nothing if the value is already set, but then I thought it might be a good defense against reasoning errors to set it anyway: I think it shouldn't affect performance and it will raise an exception if an inconsistent internal state is reached.

Reply:
Would be good to  add that reasoning as a comment

Reply:
Done.